File: content/snippets/typescript/design_patterns/chain_of_responsibility.ts
interface IHandler {
    setNext(next: IHandler): void;
    handle(request: Request): void;
}

class Request {
    constructor(public level: number, public message: string) {}
}

abstract class BaseHandler implements IHandler {
    protected next: IHandler | null = null;
    
    setNext(next: IHandler): void {
        this.next = next;
    }
    
    abstract handle(request: Request): void;
}

class LowHandler extends BaseHandler {
    handle(request: Request): void {
        if (request.level <= 1) {
            console.log(`LowHandler: ${request.message}`);
        } else if (this.next) {
            this.next.handle(request);
        }
    }
}

class MidHandler extends BaseHandler {
    handle(request: Request): void {
        if (request.level === 2) {
            console.log(`MidHandler: ${request.message}`);
        } else if (this.next) {
            this.next.handle(request);
        }
    }
}

class HighHandler extends BaseHandler {
    handle(request: Request): void {
        if (request.level >= 3) {
            console.log(`HighHandler: ${request.message}`);
        }
    }
}

const low = new LowHandler();
const mid = new MidHandler();
const high = new HighHandler();

low.setNext(mid);
mid.setNext(high);

low.handle(new Request(1, "Simple"));
low.handle(new Request(2, "Normal"));
low.handle(new Request(3, "Critical"));

File: content/snippets/python/design_patterns/chain_of_responsibility.py
from abc import ABC, abstractmethod
from typing import Optional

class Request:
    def __init__(self, level: int, message: str):
        self.level = level
        self.message = message

class Handler(ABC):
    def __init__(self):
        self.next: Optional[Handler] = None
    
    def set_next(self, handler: "Handler") -> "Handler":
        self.next = handler
        return handler
    
    @abstractmethod
    def handle(self, request: Request) -> str:
        pass

class LowHandler(Handler):
    def handle(self, request: Request) -> str:
        if request.level <= 1:
            return f"LowHandler: {request.message}"
        return self.next.handle(request) if self.next else ""

class MidHandler(Handler):
    def handle(self, request: Request) -> str:
        if request.level == 2:
            return f"MidHandler: {request.message}"
        return self.next.handle(request) if self.next else ""

class HighHandler(Handler):
    def handle(self, request: Request) -> str:
        if request.level >= 3:
            return f"HighHandler: {request.message}"
        return ""

low = LowHandler()
mid = MidHandler()
high = HighHandler()

low.set_next(mid).set_next(high)

print(low.handle(Request(1, "Simple")))
print(low.handle(Request(2, "Normal")))
print(low.handle(Request(3, "Critical")))
